interactive_engine/src/lgraph/src/jna/ffi_test.cc

```diff
@@ -0,0 +1,657 @@
+  auto e2 = r2.unwrap();
+  if (e2.Valid()) {
+    dumper << "[Error] Expect null edge handle, but got an exact one!\n";
+    return false;
+  }
+  dumper << "[INFO] Got expected null edge handle.\n";
+  dumper << "[INFO] --- Case_2 Passed!\n";
+
+  return true;
+}
+
+bool TestGetKthInEdge(Snapshot* ss) {
+  // Query src vertex: <VertexID: 10454779632061085998> <Label: software> <id: 3> <name: lop> <lang: java>
+  VertexId query_vid = 10454779632061085998U;
+  EdgeRelation query_edge_rel = EdgeRelation::From(created_EdgeLabelId, person_LabelId, software_LabelId);
+  SerialId k1 = 2;
+  VertexId expect_k1_nbr_vid = 16401677891599130309U;
+  SerialId k2 = 5;
+
+  auto r1 = ss->GetKthInEdge(query_vid, query_edge_rel, k1);
+  if (r1.isErr()) {
+    dumper << "[Error] GetKthInEdge: " << r1.unwrapErr().GetInfo() << "\n";
+    return false;
+  }
+  auto e1 = r1.unwrap();
+  if (!e1.Valid()) {
+    dumper << "[Error] Got invalid edge handle at GetKthInEdge for (" << query_vid << ", created, " << k1 << ")!\n";
+    return false;
+  }
+  auto k1_nbr_vid = e1.GetEdgeId().src_vertex_id;
+  if (k1_nbr_vid != expect_k1_nbr_vid) {
+    dumper << "[Error] The k-th in-neighbour mismatched!(k = " << k1 << ") "
+           << "Expect VertexId: " << expect_k1_nbr_vid << ", Got VertexId: " << k1_nbr_vid << "!\n";
+    return false;
+  }
+  dumper << "[INFO] Got k-th 'created' in neighbour: " << k1_nbr_vid << "\n";
+  dumper << "[INFO] --- Case_1 Passed!\n";
+
+  auto r2 = ss->GetKthInEdge(query_vid, query_edge_rel, k2);
+  if (r2.isErr()) {
+    dumper << "[Error] GetKthInEdge: " << r2.unwrapErr().GetInfo() << "\n";
+    return false;
+  }
+  auto e2 = r2.unwrap();
+  if (e2.Valid()) {
+    dumper << "[Error] Expect null edge handle, but got an exact one!\n";
+    return false;
+  }
+  dumper << "[INFO] Got expected null edge handle.\n";
+  dumper << "[INFO] --- Case_2 Passed!\n";
+
+  return true;
+}
+
+bool TestGetSnapshotId(Snapshot* ss) {
+  SnapshotId expect_ss_id = std::numeric_limits<uint32_t>::max();
+  auto ss_id = ss->GetSnapshotId();
+  if (ss_id != expect_ss_id) {
+    dumper << "[Error] Snapshot id mismatched! "
+           << "Expect: " << expect_ss_id << ", Got: " << ss_id << "!\n";
+    return false;
+  }
+  dumper << "[INFO] Got correct snapshot id: " << ss_id << "\n";
+
+  return true;
+}
+
+const unsigned test_num = 11;
+
+typedef bool (*TestFunc)(Snapshot* ss);
+
+bool RunTest(unsigned id, const std::string& test_name, TestFunc f, Snapshot* ss) {
+  dumper << "[INFO] ----------------------------------------------\n";
+  dumper << "[INFO] --- " << test_name << " Test [" << id << "/" << test_num << "]\n";
+  dumper << "[INFO] ----------------------------------------------\n";
+  bool result = f(ss);
+  dumper << "[INFO] --- " << (result? "PASSED!" : "FAILED!") << "\n";
+  dumper << "[INFO]\n";
+  return result;
+}
+
+bool runLocalTests() {
+  assert(local_graph_handle_ != nullptr);
+
+  dumper.open("./target/surefire-reports/lgraph_ffi_test.dump");
```

Comment:
Don't use hard-code path. Don't use path depend on surefire plugin.

Better just return the string to avoid writing file. Or at least get the file path from the function paramter.

Reply:
Returning the string is not an elegant way, as Java need to release the string memory allocated by C, or maybe returning the file path is better